core/src/main/java/org/geysermc/geyser/registry/populator/CustomBlockRegistryPopulator.java

```diff
@@ -0,0 +1,451 @@
+                int id = BlockRegistries.JAVA_IDENTIFIER_TO_ID.getOrDefault(javaIdentifier, -1);
+                blockStateOverrides.put(id, customBlockState.state());
+                BoxComponent extendedCollisionBox = customBlockState.extendedCollisionBox();
+                if (extendedCollisionBox != null) {
+                    CustomBlockData extendedCollisionBlock = extendedCollisionBoxSet.computeIfAbsent(extendedCollisionBox, box -> {
+                        CustomBlockData collisionBlock = createExtendedCollisionBlock(box, extendedCollisionBoxSet.size());
+                        customBlocks.add(collisionBlock);
+                        return collisionBlock;
+                    });
+                    extendedCollisionBoxes.computeIfAbsent(extendedCollisionBlock, k -> new HashSet<>())
+                            .add(id);
+                }
+            });
+        });
+    
+        BlockRegistries.CUSTOM_BLOCK_STATE_OVERRIDES.set(blockStateOverrides);
+        GeyserImpl.getInstance().getLogger().info("Registered " + blockStateOverrides.size() + " custom block overrides.");
+
+        BlockRegistries.CUSTOM_BLOCK_ITEM_OVERRIDES.set(customBlockItemOverrides);
+        GeyserImpl.getInstance().getLogger().info("Registered " + customBlockItemOverrides.size() + " custom block item overrides.");
+
+        BlockRegistries.EXTENDED_COLLISION_BOXES.set(extendedCollisionBoxes);
+        GeyserImpl.getInstance().getLogger().info("Registered " + extendedCollisionBoxes.size() + " custom block extended collision boxes.");
+    }
+
+    /**
+     * Registers all non-vanilla custom blocks defined by API
+     */
+    private static void populateNonVanilla() {
+        BlockRegistries.NON_VANILLA_BLOCK_STATE_OVERRIDES.set(nonVanillaBlockStateOverrides);
+        GeyserImpl.getInstance().getLogger().info("Registered " + nonVanillaBlockStateOverrides.size() + " non-vanilla block overrides.");
+    }
+
+    /**
+     * Registers all bedrock custom blocks defined in previous stages
+     */
+    private static void registration() {
+        BlockRegistries.CUSTOM_BLOCKS.set(customBlocks.toArray(new CustomBlockData[0]));
+        GeyserImpl.getInstance().getLogger().info("Registered " + customBlocks.size() + " custom blocks.");
+    }
+
+    /**
+     * Generates and appends all custom block states to the provided list of custom block states
+     * Appends the custom block states to the provided list of NBT maps
+     * @param customBlock the custom block data to generate states for
+     * @param blockStates the list of NBT maps to append the custom block states to
+     * @param customExtBlockStates the list of custom block states to append the custom block states to 
+     * @param stateVersion the state version to use for the custom block states
+     */
+    static void generateCustomBlockStates(CustomBlockData customBlock, List<NbtMap> blockStates, List<CustomBlockState> customExtBlockStates, int stateVersion) {
```

Comment:
Can we get a access modifier explicitly set here?

Reply:
https://github.com/GeyserMC/Geyser/pull/3505#discussion_r1294145623